### PR TITLE
Don't set a default value for -o flag and fix `ODO_LOG_LEVEL` check

### PR DESCRIPTION
**Pull request by johnmcollier**
Signed-off-by: John Collier <John.J.Collier@ibm.com>

**What type of PR is this?**
> Uncomment only one ` /kind` line, and delete the rest.
> For example, `> /kind bug` would simply become: `/kind bug`

/kind bug

**What does does this PR do / why we need it**:
Removes the default value of `json` for the `-o` / `--output` flag in odo. Since our default output is not in json, we don't need to be setting the default value for that flag to that.

Previously, it was being set and causing the `ODO_LOG_LEVEL` env var not to work, due to the check here:

https://github.com/openshift/odo/blob/master/cmd/odo/odo.go#L54

See the discussion in https://github.com/openshift/odo/pull/3115

**Which issue(s) this PR fixes**:

No issue, but I can open one.

**How to test changes / Special notes to the reviewer**:
```
Johns-MacBook-Pro-3:odo johncollier$ ODO_LOG_LEVEL=4 ./odo catalog list components
I0507 22:37:13.433129   65620 preference.go:165] The path for preference file is /Users/johncollier/.odo/preference.yaml
I0507 22:37:13.433317   65620 preference.go:165] The path for preference file is /Users/johncollier/.odo/preference.yaml
I0507 22:37:14.387035   65620 preference.go:165] The path for preference file is /Users/johncollier/.odo/preference.yaml
Odo OpenShift Components:
NAME              PROJECT       TAGS                        SUPPORTED
java              openshift     11,8,latest                 YES
nodejs            openshift     10-SCL,8,8-RHOAR,latest     YES
dotnet            openshift     2.1,2.2,3.0,latest          NO
golang            openshift     1.11.5,latest               NO
httpd             openshift     2.4,latest                  NO
modern-webapp     openshift     10.x,latest                 NO
nginx             openshift     1.10,1.12,latest            NO
perl              openshift     5.24,5.26,latest            NO
php               openshift     7.0,7.1,7.2,latest          NO
python            openshift     2.7,3.6,latest              NO
ruby              openshift     2.4,2.5,latest              NO

Odo Devfile Components:
NAME                 DESCRIPTION                           SUPPORTED
maven                Upstream Maven and OpenJDK 11         YES
nodejs               Stack with NodeJS 10                  YES
openLiberty          Open Liberty microservice in Java     YES
java-spring-boot     Spring Boot® using Java               YES

```

```
Johns-MacBook-Pro-3:odo johncollier$ ./odo catalog list components -o json
{
    "kind": "List",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "items": [
        {
            "kind": "ComponentType",
            "apiVersion": "odo.openshift.io/v1alpha1",
            "metadata": {
                "name": "dotnet",
                "namespace": "openshift",
                "creationTimestamp": null
            },
            "spec": {
                "allTags": [
                    "2.1",
                    "2.2",
                    "3.0",
                    "latest"
                ],
                "nonHiddenTags": [
                    "2.1",
                    "2.2",
                    "3.0",
                    "latest"
                ],
                "supportedTags": []
            }
        },
```

```
Johns-MacBook-Pro-3:odo johncollier$ ./odo catalog list components -o yaml
 ✗  Please input a valid output format for -o, available format: json
```

### CODE REVIEW

**Comment by amitkrout** [comment]
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/3121/pull-ci-openshift-odo-master-v4.3-integration-e2e/165#1:build-log.txt%3A390

/retest

**Comment by amitkrout** [comment]
@johnmcollier It seems your change causes one spec failure on both travis and OpenShift CI. PTAL

Travis - https://travis-ci.com/github/openshift/odo/jobs/329756324#L467

**Comment by johnmcollier** [comment]
@amitkrout So I did some digging:

It appears the issue is related to retrieving the value for the `-o json`, `jsonFlagValue: 

https://github.com/openshift/odo/blob/master/cmd/odo/odo.go#L53

Basically, it seems that we only ever retrieve an "accurate" value for the `-o` flag if it immediately follows `odo` itself:
```
Johns-MacBook-Pro-3:odo johncollier$ ODO_LOG_LEVEL=4 ./odo -o json project create tester
-o flag value: json
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "unable to create new project: unable to create new project tester: project.project.openshift.io \"tester\" already exists"
}
```

But if it follows at the end of the command, we're unable to look up the proper value
```
Johns-MacBook-Pro-3:odo johncollier$ ODO_LOG_LEVEL=4 ./odo project create tester -o json
-o flag value: 
I0508 10:40:35.598886   67965 preference.go:165] The path for preference file is /Users/johncollier/.odo/preference.yaml
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "unable to create new project: unable to create new project tester: project.project.openshift.io \"tester\" already exists"
}
```

So basically, we're never actually looking up the proper value of the `-o` flag in odo.go (but we appear to be elsewhere in the code). The reason this wasn't seen before is that before my PR, the default value of the `-o` flag was `json`, so it wouldn't have been noticed.

**Comment by johnmcollier** [comment]
@amitkrout From looking through the code, I've found a fix and will be pushing it up to this PR!

**Comment by girishramnani** [comment]
/hold
This is a big decision and might affect vs-code adapter, also might require docs change. and we might need to consider the fact that the verbase logs are not written to stdout ( if I remember correctly ) so can't we just print the logs if someone provides `-v` with a json output?

**Comment by johnmcollier** [comment]
@girishramnani as far I understand things, this doesn’t impact the functionality of odo. The previous default value of “json” for `-o` didn’t seem to work (or do anything), as json output wasn’t displayed by default.

**Comment by kadel** [comment]
> /hold
> This is a big decision and might affect vs-code adapter, also might require docs change. and we might need to consider the fact that the verbase logs are not written to stdout ( if I remember correctly ) so can't we just print the logs if someone provides `-v` with a json output?

How does it affect vs-code adapter? This just fixes something that was broken. No behavior is modified.

File: cmd/odo/odo.go [hunk 1]
```diff
@@ -49,9 +48,11 @@ func main() {
...
 	// The "-v" flag set on command line will take precedence over ODO_LOG_LEVEL env
 	v := flag.CommandLine.Lookup("v").Value.String()
-	// if the json flag is passed and is valid, we don't turn on ODO_LOG_LEVEL
-	jsonFlagValue := flag.CommandLine.Lookup("o").Value.String()
-	if level, ok := os.LookupEnv("ODO_LOG_LEVEL"); ok && v == "0" && strings.ToLower(jsonFlagValue) != "json" {
+
+	// if the output flag is set, we don't turn on ODO_LOG_LEVEL
+	outputFlag := pflag.Lookup("o")
+	hasFlagChanged := outputFlag != nil && outputFlag.Changed
+	if level, ok := os.LookupEnv("ODO_LOG_LEVEL"); ok && v == "0" && hasFlagChanged {
```
Comment (kadel) [inline cmd/odo/odo.go:55]: Should the last condition be `!hasFlagChanged`?

Comment (johnmcollier) [reply cmd/odo/odo.go:55]: Ah, you're right! I totally missed the `!=` in the `strings.ToLower(jsonFlagValue) != "json"`

**Comment by johnmcollier** [comment]
So, after a few hours of debugging, I've finally found a fix for this issue that doesn't break either the `-o` flag and also fixes the `ODO_LOG_LEVEL` environment variable.

So it appears that there isn't a way to retrieve the value of the `-o` flag from within odo.go, **unless**  the flag immediately follows `odo`. I.e:

With `-o` after `odo`
```
Johns-MacBook-Pro-3:~ johncollier$ odo -o json project create tester
-o flag value in odo.go: json
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "unable to create new project: unable to create new project tester: project.project.openshift.io \"tester\" already exists"
}
```

And with `-o` at the end of the command:
```
Johns-MacBook-Pro-3:~ johncollier$ odo project create tester -o json
-o flag value in odo.go: 
I0511 16:35:52.169117   91077 preference.go:165] The path for preference file is /Users/johncollier/.odo/preference.yaml
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "unable to create new project: unable to create new project tester: project.project.openshift.io \"tester\" already exists"
}
```

This is because the other commands' flags/args have yet to be parsed, so [odo.go](https://github.com/openshift/odo/blob/80472ead1ec1c8155da274cfe8aa110ebcc3571f/cmd/odo/odo.go) can only see the flags that immediately follow the root command.

However, if we move the check for `ODO_LOG_LEVEL` outside of odo.go to somewhere the flags for all of the subcommands have already been parsed, then this issue wouldn't occur. So, what I did was move the check for `ODO_LOG_LEVEL`  to [genericclioptions.CheckMachineReadableOutputCommand](https://github.com/openshift/odo/blob/master/pkg/odo/genericclioptions/runnable.go#L67)). This function already does some validation over whether or not verbose and JSON logging is enabled, so it was easy to drop right in.

**Comment by johnmcollier** [comment]
With my new change:
```
Johns-MacBook-Pro-3:~ johncollier$ odo list
 ✗  Please specify the application name and project name
Or use the command from inside a directory containing an odo component.
Johns-MacBook-Pro-3:~ johncollier$ odo list -v 4
I0511 17:11:40.125767   41245 list.go:51] New Context
 ✗  Please specify the application name and project name
Or use the command from inside a directory containing an odo component.
Johns-MacBook-Pro-3:~ johncollier$ odo list -o json
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "Please specify the application name and project name\nOr use the command from inside a directory containing an odo component."
}
Johns-MacBook-Pro-3:~ johncollier$ odo list -o json -v 4
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "Please specify the application name and project name\nOr use the command from inside a directory containing an odo component."
}
Johns-MacBook-Pro-3:~ johncollier$ ODO_LOG_LEVEL=4 odo list
I0511 17:12:03.575813   41254 list.go:51] New Context
 ✗  Please specify the application name and project name
Or use the command from inside a directory containing an odo component.
Johns-MacBook-Pro-3:~ johncollier$ ODO_LOG_LEVEL=4 odo list -o json
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "Please specify the application name and project name\nOr use the command from inside a directory containing an odo component."
}
Johns-MacBook-Pro-3:~ johncollier$ ODO_LOG_LEVEL=4 odo list -o json -v 4
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "Please specify the application name and project name\nOr use the command from inside a directory containing an odo component."
}
```

**Comment by kadel** [comment]
/hold cancel 
This was discussed in odo contributor call. @girishramnani is ok with that. And I've checked and verified that vscode plugin is calling odo using full `-o json` everywhere.

**Comment by kadel** [comment]
/hold cancel 
This was discussed in odo contributor call. @girishramnani is ok with that. And I've checked and verified that vscode plugin is calling odo using full `-o json` everywhere.
